### ISSUE TITLE
logging: how to make a Logger the default std logger?

### ISSUE THREAD

**Issue by mpl** [GoogleCloudPlatform/google-cloud-go#439]
Hi,

With the V1 logging API, there was an easy way not only to make a Google Cloud Logger the default one, e.g.

`
logc, _ := logging.NewClient(context.Background(), projID, "myprojectstderr")
log.SetOutput(logc.Writer(logging.Debug))
`

but even better, to let the default logger write to both Google Cloud Logger, and say, stderr, at the same time:

`
logc, _ := logging.NewClient(context.Background(), projID, "myprojectstderr")
log.SetOutput(io.MultiWriter(os.Stderr, logc.Writer(logging.Debug)))
`

I've noticed that the Writer method is gone from the V2 API, and I haven't found any obvious replacement.

What is now the idiomatic way to do something like the above with the V2?

Regards,
Mathieu

**Comment by jba** [comment]
There isn't a way. But if you have a logger, you could write your own io.Writer implementation quite easily:
```
type writer struct {
    severity logging.Severity
    logger   *logging.Logger
}

func (w writer) Write(p []byte) (n int, err error) {
    w.logger.Log(logging.Entry{
        Severity: w.severity,
        Payload:  string(p),
    })
    return len(p), nil
}
```

**Comment by mpl** [comment]
Thanks.
I was wondering about that, and I thought automatically creating one log entry per write call seemed overkill/too much.
I admit I haven't checked the source code for how the V1 Writer worked, but I had assumed it was smarter than that, with maybe some buffering. But it's true I don't see what kind of heuristics it could have used to decide when's a good time (i.e. after how many Write calls) to finally actually create and write a log entry.

**Comment by jba** [comment]
The documentation on `log.Logger` says
> Each logging operation makes a single call to the Writer's Write method.

So one entry per call to Write makes sense.

Entries are cheap to create and the Log method is asynchronous, with lots of buffering, so performance should be OK.

You could also have your writer call LogSync, but that would be slow.

**Comment by mpl** [comment]
Ok.

What's the rationale for not having kept a Writer method which would do something like what you gave as an example? Granted it is easy for oneself to write their own, but it seems useful enough (imo) to warrant a default implementation provided by the package, so that users don't even have to wonder about it. Especially since the V1 was providing one.

**Comment by jba** [comment]
The bottom line is that io.Writer is too low-level an interface for logging. It happens to work for the standard log package, because of the guarantee I quoted above. But if the API provided an io.Writer, people might use it in many of the places where io.Writers normally work, often with surprising results (entries that are binary, entries that are sliced at 8K boundaries, etc.) The right interface to a logger would be at the entry level, but the standard log package doesn't provide that.

We did want to make it easy to use the standard log package, so we provided a log.Logger for each severity. But we weren't willing to export the writer, given the opportunity for confusion.
